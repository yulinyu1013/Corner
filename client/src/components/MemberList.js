import React, { useState, useEffect } from 'react';
import { message } from 'antd';
import '../styles/memberlist.css';
import 'antd/dist/antd.css';
import { FaSearch } from 'react-icons/fa';
import {
  getMemberList,
  getCurrentUserLevel,
  setRoleInCorner,
  removeUserFromCorner,
} from '../fetchers/userApis';
import Member from './Member';

function MemberList({
  isLoggedIn,
  currentUser,
  cornerId,
  isChat,
}) {
  const [memberList, setMemberList] = useState();
  const [searchTerm, setSearchTerm] = useState('');
  const [currentUserLevel, setCurrentUserLevel] = useState({
    id: 0,
    name: currentUser.name,
    avatar: null,
    level: -1,
  });

  useEffect(() => {
    console.log(isLoggedIn);
    // if loggedIn, fetch current user's level in the corner and etc.
    if (isLoggedIn) {
      // currentUser.id = 'test1';
      console.log('-------');
      console.log(currentUser.id);
      getCurrentUserLevel(currentUser.id, cornerId).then((res) => {
        console.log('-------');
        console.log(res);
        setCurrentUserLevel(res);
      });
    }
    // fetch the current memberlist
    getMemberList(cornerId).then((res) => {
      setMemberList(res);
      console.log(res);
    });
  }, [isLoggedIn, currentUser]);

  useEffect(() => {
    if (searchTerm !== '') {
      const lowSearchTerm = searchTerm.toLowerCase();
      const filteredData = [...memberList].filter(
        (val) => val.name.toLowerCase().includes(lowSearchTerm),
      );
      setMemberList(filteredData);
    } else {
      // fetch the current memberlist
      getMemberList(cornerId).then((res) => {
        setMemberList(res);
      });
    }
  }, [searchTerm, isLoggedIn]);

  const deleteUser = (id) => {
    if (currentUserLevel.level === 1) {
      message.info('Sorry! You are not authorized to delete a user');
      return;
    }
    if (id === currentUser.id) {
      message.info('Sorry! You cannot delete yosurself');
      return;
    }
    const filteredData = [...memberList].filter(
      (val) => val.id !== id,
    );
    const newList = filteredData.sort((m1, m2) => (m2.level - m1.level));
    setMemberList(newList);
    removeUserFromCorner(cornerId, id);
  };

  const setAdmin = (id) => {
    // console.log(currentUser);
    if (currentUserLevel.level === 1) {
      message.info('Sorry! You are not authorized to set an admin role');
      return;
    }
    const newAdmin = [...memberList].filter(
      (val) => val.id === id,
    );
    const temp = [...memberList].filter(
      (val) => val.id !== id,
    );
    newAdmin[0].level = 2;
    const newList = temp.concat(newAdmin).sort((m1, m2) => (m2.level - m1.level));
    setMemberList(newList);
    setRoleInCorner(id, cornerId, 2);
  };

  const revokeAdmin = (id) => {
    if (currentUserLevel.level === 1) {
      message.info('Sorry! You are not authorized to revoke an admin role');
      return;
    }
    const newAdmin = [...memberList].filter(
      (val) => val.id === id,
    );
    const temp = [...memberList].filter(
      (val) => val.id !== id,
    );
    newAdmin[0].level = 1;
    const newList = temp.concat(newAdmin).sort((m1, m2) => (m2.level - m1.level));
    setMemberList(newList);
    setRoleInCorner(id, cornerId, 1);
  };

  return (
    <div className="member-list-outer">
      {memberList && (
        <div className="member-list-wrapper">
          <div className="member-header">
            <span>Members</span>
          </div>
          <div className="search-bar">
            <FaSearch className="search-icon" />
            <input
              className="input-field"
              type="text"
              placeholder="Search member..."
              onChange={(e) => {
                setSearchTerm(e.target.value);
              }}
            />
          </div>
          <div className="list-container">
            {memberList.map((user) => (
              <Member
                currentUser={currentUserLevel}
                user={user}
                deleteUser={deleteUser}
                setAdmin={setAdmin}
                revokeAdmin={revokeAdmin}
                isLoggedIn={isLoggedIn}
                isChat={isChat}
              />
            ))}
          </div>
        </div>
      )}
    </div>
  );
}

export default MemberList;
