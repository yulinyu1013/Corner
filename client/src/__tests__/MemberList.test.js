import React from 'react';
import renderer from 'react-test-renderer';
import MemberList from '../components/MemberList';

const user = {
  id: 0,
  name: 'dummy',
  avatar: null,
  level: -1,
}

it('renders MemberList correctly when user loggedin and currentUser not empty', () => {
  const tree = renderer.create(<MemberList isLoggedIn currentUser={user} cornerId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders MemberList correctly when user not and currentUser is empty', () => {
  const tree = renderer.create(<MemberList isLoggedIn={false} currentUser={user} cornerId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});
