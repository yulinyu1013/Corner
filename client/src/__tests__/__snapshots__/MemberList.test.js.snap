// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MemberList correctly when user loggedin and currentUser not empty 1`] = `
<div
  className="member-list-outer"
/>
`;

exports[`renders MemberList correctly when user not and currentUser is empty 1`] = `
<div
  className="member-list-outer"
/>
`;
